odl/space/base_tensors.py

```diff
@@ -0,0 +1,741 @@
+                if self.__complex_space is None:
+                    self.__complex_space = self._astype(dtype, self.order)
+                return self.__complex_space
+            else:
+                return self._astype(dtype, order)
+        else:
+            return self._astype(dtype, order)
+
+    @property
+    def order(self):
+        """Guaranteed data storage order in this space.
+
+        This is one of ``('A', 'C', 'F')``, where ``'A'`` means "any",
+        i.e. "no guarantee".
+        """
+        return self.__order
+
+    @property
+    def new_elem_order(self):
+        """Storage order for new elements in this space.
+
+        This is identical to `order` except for ``self.order == 'A'``,
+        where ``'C'`` is returned.
+        """
+        return 'C' if self.order == 'A' else self.order
+
+    @property
+    def size(self):
+        """Total number of entries in an element of this space."""
+        return np.prod(self.shape)
+
+    @property
+    def ndim(self):
+        """Number of axes (=dimensions) of this space, also called "rank"."""
+        return len(self.shape)
+
+    def __len__(self):
+        """Number of tensor entries along the first axis."""
+        return self.shape[0]
+
+    @property
+    def itemsize(self):
+        """Size in bytes of one entry in an element of this space."""
+        return self.dtype.itemsize
+
+    @property
+    def nbytes(self):
+        """Total number of bytes used by an element of this space in memory."""
+        return self.size * self.itemsize
+
+    def __contains__(self, other):
+        """Return ``other in self``.
+
+        Returns
+        -------
+        contains : bool
+            ``True`` if ``other`` has a ``space`` attribute that is equal
+            to this space, ``False`` otherwise.
+
+        Examples
+        --------
+        Elements created with the `TensorSpace.element` method are
+        guaranteed to be contained in the same space:
+
+        >>> spc = odl.tensor_space((2, 3), dtype='uint64')
+        >>> spc.element() in spc
+        True
+        >>> x = spc.element([[0, 1, 2],
+        ...                  [3, 4, 5]])
+        >>> x in spc
+        True
+
+        Sizes, data types and other essential properties characterize
+        spaces and decide about membership:
+
+        >>> smaller_spc = odl.tensor_space((2, 2), dtype='uint64')
+        >>> y = smaller_spc.element([[0, 1],
+        ...                          [2, 3]])
+        >>> y in spc
+        False
+        >>> x in smaller_spc
+        False
+        >>> other_dtype_spc = odl.tensor_space((2, 3), dtype='uint32')
+        >>> z = other_dtype_spc.element([[0, 1, 2],
+        ...                              [3, 4, 5]])
+        >>> z in spc
+        False
+        >>> x in other_dtype_spc
+        False
+
+        On the other hand, spaces are not unique:
+
+        >>> spc2 = odl.tensor_space((2, 3), dtype='uint64')  # same as spc
```

Comment:
I actually changed them all on the GPU branch to simple shape, will do something similar here.